File: Week 1/cpp/15.typecasting.cpp
/*
 * WEEK 1 - C++ FUNDAMENTALS
 * Topic: Typecasting (Implicit and Explicit Conversions)
 * File: 15.typecasting.cpp
 *
 * CONCEPT:
 * C++ has two flavours of casts: IMPLICIT (compiler does it for you, e.g.
 * widening int -> long -> double) and EXPLICIT (you write a cast). Modern
 * C++ favours the four "named" casts:
 *   static_cast<T>(x)       - safe value-changing conversion
 *   const_cast<T>(x)         - add/remove const qualifier
 *   reinterpret_cast<T>(x)  - low-level bit reinterpretation (DANGEROUS)
 *   dynamic_cast<T>(x)       - safe downcast in polymorphic hierarchies
 *
 * KEY POINTS:
 *  - Widening (int -> long, int -> double) is implicit.
 *  - Narrowing (double -> int) requires an explicit cast; truncates toward zero.
 *  - C-style cast `(T) x` still works but is discouraged in new code.
 *  - `static_cast<long>(d)` matches Java's `(long) d`.
 *  - Brace initialisation `int n{d};` REJECTS narrowing at compile time.
 *
 * SYNTAX:
 *   long  l1 = i;                       // implicit widening
 *   long  l2 = static_cast<long>(d);    // explicit narrowing (truncates)
 *   int   n  = (int) 3.7;               // C-style cast (still legal)
 *
 * DRY RUN:
 *  i=100;       l1 = i               -> 100
 *  d=100.04;    l2 = static_cast<long>(d) -> 100  (decimal truncated)
 *  d=-100.99;   static_cast<long>(d) -> -100 (NOT -101)
 */

#include <iostream>

int main() {
    int i = 100;
    long l1 = i;                                  // implicit widening
    double d = 100.04;
    long l2 = static_cast<long>(d);              // explicit narrowing
    long l3 = (long)d;                            // C-style equivalent

    std::cout << i << '\n';
    std::cout << l1 << '\n';
    std::cout << d << '\n';
    std::cout << l2 << "  (static_cast)\n";
    std::cout << l3 << "  (C-style cast)\n";

    // Truncation (toward zero) demo
    std::cout << "static_cast<int>(-100.99) = "
              << static_cast<int>(-100.99) << "  (NOT -101)\n";

    // Narrowing rejection with braces
    // int bad{d};   // compile error: narrowing conversion from double to int

    // Implicit promotion happens during arithmetic
    int    a = 1;
    double b = 2.5;
    auto   sum = a + b;            // double, value 3.5
    std::cout << "1 + 2.5 = " << sum << " (type deduced via auto)\n";

    return 0;
}

/*
 * NOTES:
 *  - Java forbids implicit narrowing; C++ permits it via implicit conversions
 *    (with warnings). Use brace-init for safety.
 *  - The four named casts make the intent of each conversion explicit and
 *    grep-able — preferred over the bare `(T) x` C-style cast.
 *  - For converting between pointer/reference types in class hierarchies,
 *    use dynamic_cast (Java equivalent: `(SubClass) obj` with `instanceof`).
 */

File: Week 1/python/15.typecasting.py
"""
WEEK 1 - PYTHON FUNDAMENTALS
Topic: Typecasting (Type Conversion)
File: 15.typecasting.py

CONCEPT:
Type conversion in Python is done with the constructor of the target type:
`int(x)`, `float(x)`, `str(x)`, `bool(x)`, etc. Python does NOT distinguish
"widening" vs "narrowing" the way Java does — every conversion is explicit
and uses the same syntax.

KEY POINTS:
- `int(3.7)`     -> 3   (truncates toward zero, matches Java's `(int) 3.7`)
- `int(-3.7)`    -> -3  (truncate toward zero, NOT floor)
- `float(5)`     -> 5.0 (Java: `(double) 5` or implicit widening)
- `str(42)`      -> "42" (Java: `Integer.toString(42)` or `"" + 42`)
- `int("123")`   -> 123 (Java: `Integer.parseInt("123")`)
- `bool(0)`      -> False; `bool(1)` -> True; `bool("")` -> False
- Implicit "widening" int -> float happens only inside arithmetic (1 + 2.0 -> 3.0)

PYTHON-SPECIFIC NOTES vs Java:
- No casting syntax `(type)value` — always a function call
- `int(s)` raises `ValueError` for bad strings (Java throws NumberFormatException)
- `int(x, base)` parses strings in any base 2..36

DRY RUN:
1. i = 100;  l1 = i             -> 100   (already int)
2. d = 100.04; l2 = int(d)       -> 100   (truncates)
3. int("1010", 2)                -> 10    (binary parse)
4. float("3.14e2")               -> 314.0
"""


def main() -> None:
    # --- "Widening" int -> float (implicit during arithmetic) ---
    i = 100
    f = i + 0.0          # implicit promotion to float
    print(f"i + 0.0 = {f}  ({type(f).__name__})")

    # --- "Narrowing" float -> int (explicit, truncates toward zero) ---
    d = 100.04
    truncated = int(d)
    print(f"int({d}) = {truncated}        # truncates the decimal part")

    print(f"int(-100.99) = {int(-100.99)}  # -100, NOT -101 (truncates toward zero)")
    print(f"floor(-100.99) = {-101}        # use math.floor for floor semantics")

    # --- str <-> numeric ---
    print(f"str(42)       = {str(42)!r}")
    print(f"int('123')    = {int('123')}")
    print(f"float('3.14') = {float('3.14')}")
    # Bases — handy for binary, octal, hex
    print(f"int('1010', 2) = {int('1010', 2)}    # binary")
    print(f"int('1A', 16) = {int('1A', 16)}     # hex")

    # --- bool conversions ---
    print(f"bool(0)   = {bool(0)}")
    print(f"bool(1)   = {bool(1)}")
    print(f"bool('')  = {bool('')}")
    print(f"bool([])  = {bool([])}")
    print(f"bool('x') = {bool('x')}")

    # --- Java demo equivalent ---
    print("\n=== Java demo equivalent ===")
    java_i = 100
    java_l1 = java_i              # widening (int -> long is automatic in Java; here just int)
    java_d = 100.04
    java_l2 = int(java_d)         # explicit narrowing
    print(java_i)                # 100
    print(java_l1)               # 100
    print(java_d)                # 100.04
    print(java_l2)               # 100


if __name__ == "__main__":
    main()


# NOTES:
# - Java has TWO categories: widening (automatic) and narrowing (explicit cast).
#   Python ALWAYS requires explicit conversion via constructor functions.
# - `int(x)` truncates toward zero; `math.floor(x)` floors toward -infinity.
# - Strings -> int: invalid input raises ValueError. Wrap with try/except for safety.
# - Python's `bool` is a subclass of int — `True == 1`, `False == 0`.
